File: backend/auth/dependencies.py
from fastapi import Depends, HTTPException, status, Request
from fastapi.security import OAuth2PasswordBearer

# This is now mainly for documentation and OpenAPI spec generation
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

def get_current_user(request: Request) -> dict:
    """
    Dependency that retrieves the current user from the request state,
    which is populated by the FirebaseAuthMiddleware.
    
    Raises:
        HTTPException: If no authenticated user is found in the request state.
    """
    if not hasattr(request.state, "user") or not request.state.user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Not authenticated",
            headers={"WWW-Authenticate": "Bearer"},
        )
    return request.state.user

File: backend/auth/dependencies.js
/**
 * Authentication dependencies for route protection
 */

/**
 * Middleware that ensures the request has an authenticated user.
 * Must be used after firebaseAuthMiddleware.
 */
function requireAuth(req, res, next) {
    if (!req.user) {
        return res.status(401).json({
            detail: 'Not authenticated',
            headers: { 'WWW-Authenticate': 'Bearer' }
        });
    }
    next();
}

/**
 * Helper function to get current user from request.
 * Throws if no user is authenticated.
 * @param {Request} req - Express request object
 * @returns {object} The authenticated user's decoded token
 */
function getCurrentUser(req) {
    if (!req.user) {
        const error = new Error('Not authenticated');
        error.status = 401;
        throw error;
    }
    return req.user;
}

module.exports = {
    requireAuth,
    getCurrentUser
};
